### PR TITLE
fix: Avoid race condition in Java runtime sequence concatenation optimization

### LINKED ISSUE #1454: Lazy sequence concatenation in C# and Java runtimes is not thread-safe
Both [C#](https://github.com/dafny-lang/dafny/blob/90eb1084d9862384accd03fb5a1c8af500212a05/Source/DafnyRuntime/DafnyRuntime.cs#L1111-L1120) and [Java](https://github.com/dafny-lang/dafny/blob/90eb1084d9862384accd03fb5a1c8af500212a05/Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java#L756-L761) optimize sequence concatenation by creating a lazy representation with internal mutable state. I've suspected this isn't thread safe and have now confirmed it isn't in C# at least, using Coyote. Namely, two threads can both determine that the lazy representation needs to be resolved, one thread will complete this operation and clear the `left` and `right` fields so they can be GC'd, and then the second thread throws a NullReferenceException when trying to do the same.

One possible answer to this is "not an issue, the runtimes and Dafny itself aren't guaranteed to be thread-safe", but I'd be personally disappointed by this given this seems to be the only such issue with the runtimes AFAICT so far. The solution in C# is likely to use the fairly sophisticated and thread-safe `ImmutableList` class instead of our hand-rolled implementation of this immutable datatype, and the Java runtime likely has something similar we can wrap.

See also https://github.com/dafny-lang/rfcs/pull/6.

### PR DESCRIPTION
**Pull request by robin-aws**
Fixes #1454.

Uses the same technique as the C# runtime, the only wrinkle being finding the left-most node in order to call `newCopied(length)` on it. This allows us to actually assert success in the multithreaded benchmark for this case.

This required being able to install the Java runtime to Maven local, so I've also applied the same changes @alex-chew had queued up for publishing this package.

Also added support for a `{:benchmarkTearDown}` attribute in the plugin, for once-per-benchmark cleanup, statistics reporting, etc.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java [hunk 1]
```diff
@@ -859,11 +879,13 @@ private NonLazyDafnySequence<T> computeElements() {
             if (seq instanceof ConcatDafnySequence<?>) {
                 ConcatDafnySequence<T> cseq = (ConcatDafnySequence<T>) seq;
 
-                if (cseq.ans != null) {
-                    copier.copyFrom(cseq.ans);
+                leftBuffer = cseq.left;
+                rightBuffer = cseq.right;
+                if (leftBuffer == null || rightBuffer == null) {
```
Comment (alex-chew) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java:884]: Isn't this condition backwards? i.e.
```suggestion
                if (leftBuffer != null && rightBuffer != null) {
```

**Review by alex-chew — CHANGES_REQUESTED** [review]

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java (hunk 1 above)
Comment (robin-aws) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java:884]: Good catch. I flipped the if and else bodies instead to stay consistent with the other cases, though.

**Review by alex-chew — APPROVED** [review]
It's unfortunate that we don't have better tooling to test concurrency-related changes like this, but otherwise this LGTM.

**Review by keyboardDrummer — APPROVED** [review]

**Comment by keyboardDrummer** [comment]
I don't follow what the Maven publishing is for. Can you elaborate?

**Pull request merged**